File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cs
public class Solution {
    public int[] SearchRange(int[] nums, int target) {
        int[] result = new int[]{-1, -1};
        
        if(nums.Length == 0) return result;
        
        int firstPos;
        int lastPos;
        
        firstPos = BinarySearch(nums, target, 'l');
        
        if(firstPos == -1) {
            return result;
        }
        
        lastPos = BinarySearch(nums, target, 'r');
        
        result[0] = firstPos;
        result[1] = lastPos;
        
        return result;
    }
    
    public int BinarySearch(int[] nums, int target, char type) {
        int left = 0;
        int right = nums.Length - 1;
        int index = -1;
        
        while(left <= right) {
            int middle = (left + right)/2;
            if(nums[middle] == target) {
                index = middle;
                if(type == 'l') {
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            }
            else if(nums[middle] < target) left = middle + 1;
            else right = middle - 1;
        }
        
        return index;
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return [-1, -1]
        
        first_position = self.binarySearch(nums, target, 'l')
        if(first_position == -1):
            return [-1, -1]
        
        last_position = self.binarySearch(nums, target, 'r')
        
        return [first_position, last_position]
        
    def binarySearch(self, nums: List[int], target: int, char_type: chr) -> int:
        left = 0
        right = len(nums) - 1
        pos = -1
        
        while(left <= right):
            mid = left + ((right - left) // 2)
            if(nums[mid] == target):
                pos = mid
                if(char_type == 'l'):
                    right = mid - 1
                else:
                    left = mid + 1
            
            elif(nums[mid] > target):
                right =  mid - 1
            else:
                left = mid + 1
        
        return pos
#do a binary search left to find the first value target in nums, if we see that target value, then move left and continue our binary search until we see the lowest index of target. if we don't see any value, return -1.
#do a binary search right to find the first value target in nums, if we see that target value, then move right and continue our binary search until we see the highest index of target. if we don't see any value, return -1.